services/user/Sites/src/Sites.cpp

```diff
@@ -105,27 +254,144 @@ namespace SystemService
+                  if (is_accepted(encoding))
+                  {
+                     headers.push_back({"Content-Encoding", encoding});
+                  }
+                  else
+                  {
+                     auto decompressor = get_decompressor(encoding);
+                     // If identity is accepted and we can decompress the content, do that
+                     // Otherwise, 406
+                     if (!is_accepted("identity") || !decompressor)
+                     {
+                        return make406("Requested encoding not supported.");
+                     }
+
+                     // Decompress the content (don't bother if it's a HEAD request)
+                     if (request.method != "HEAD")
+                     {
+                        auto decompressorAccount = AccountNumber{*decompressor};
+
+                        check(to<Accounts>().exists(decompressorAccount),
+                              "[Fallback decompression error] Decompressor account not found");
+
+                        Actor<DecompressorInterface> decoder(Sites::service, decompressorAccount);
+                        content->content = decoder.decompress(content->content);
+                     }
+                     headers.push_back({"Content-Encoding", "identity"});
+                  }
+               }
+            }
+            else  // Request does not specify accepted encodings
+            {
+               if (content->contentEncoding)
+               {
+                  headers.push_back({"Content-Encoding", *content->contentEncoding});
+               }
+               else
+               {
+                  headers.push_back({"Content-Encoding", "identity"});
+               }
+            }
+
+            auto reply = HttpReply{
                 .contentType = content->contentType,
-                .body        = content->content,
+                .headers     = std::move(headers),
             };
+
+            if (request.method != "HEAD")
+            {
+               reply.body = content->content;
+            }
+
+            return reply;
          }
       }
 
       return std::nullopt;
-   }  // Sites::serveSys
+   }
 
-   void Sites::storeSys(std::string path, std::string contentType, std::vector<char> content)
+   void Sites::storeSys(std::string                path,
+                        std::string                contentType,
+                        std::optional<std::string> contentEncoding,
+                        std::vector<char>          content)
    {
       Tables tables{};
       auto   table = tables.template open<SitesContentTable>();
 
       check(path.starts_with('/'), "Path doesn't begin with /");
+      auto hash = psio::detail::seahash(std::string_view(content.data(), content.size()));
+
+      if (contentEncoding)
+      {
+         check(std::find(validEncodings.begin(), validEncodings.end(), *contentEncoding) !=
```

Comment:
Since validEncodings is static, I would make sure it's sorted and then use `std::ranges::binary_search`.